## 1. Counting missing values
<p>Sports clothing and athleisure attire is a huge industry, worth approximately <a href="https://www.statista.com/statistics/254489/total-revenue-of-the-global-sports-apparel-market/">$193 billion in 2021</a> with a strong growth forecast over the next decade! </p>
<p>In this notebook, we play the role of a product analyst for an online sports clothing company. The company is specifically interested in how it can improve revenue. We will dive into product data such as pricing, reviews, descriptions, and ratings, as well as revenue and website traffic, to produce recommendations for its marketing and sales teams.  </p>
<p>The database provided to us, <code>sports</code>, contains five tables, with <code>product_id</code> being the primary key for all of them: </p>
<h3 id="info"><code>info</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_name</code></td>
<td><code>varchar</code></td>
<td>Name of the product</td>
</tr>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>description</code></td>
<td><code>varchar</code></td>
<td>Description of the product</td>
</tr>
</tbody>
</table>
<h3 id="finance"><code>finance</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>listing_price</code></td>
<td><code>float</code></td>
<td>Listing price for product</td>
</tr>
<tr>
<td><code>sale_price</code></td>
<td><code>float</code></td>
<td>Price of the product when on sale</td>
</tr>
<tr>
<td><code>discount</code></td>
<td><code>float</code></td>
<td>Discount, as a decimal, applied to the sale price</td>
</tr>
<tr>
<td><code>revenue</code></td>
<td><code>float</code></td>
<td>Amount of revenue generated by each product, in US dollars</td>
</tr>
</tbody>
</table>
<h3 id="reviews"><code>reviews</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_name</code></td>
<td><code>varchar</code></td>
<td>Name of the product</td>
</tr>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>rating</code></td>
<td><code>float</code></td>
<td>Product rating, scored from <code>1.0</code> to <code>5.0</code></td>
</tr>
<tr>
<td><code>reviews</code></td>
<td><code>float</code></td>
<td>Number of reviews for the product</td>
</tr>
</tbody>
</table>
<h3 id="traffic"><code>traffic</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>last_visited</code></td>
<td><code>timestamp</code></td>
<td>Date and time the product was last viewed on the website</td>
</tr>
</tbody>
</table>
<h3 id="brands"><code>brands</code></h3>
<table>
<thead>
<tr>
<th>column</th>
<th>data type</th>
<th>description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>product_id</code></td>
<td><code>varchar</code></td>
<td>Unique ID for product</td>
</tr>
<tr>
<td><code>brand</code></td>
<td><code>varchar</code></td>
<td>Brand of the product</td>
</tr>
</tbody>
</table>
<p>We will be dealing with missing data as well as numeric, string, and timestamp data types to draw insights about the products in the online store. Let's start by finding out how complete the data is.</p>

In [650]:
%%sql
postgresql:///sports 
select 
        count(*) as total_rows 
      -,count(distinct a.product_id)  as distc_product
       ,count(case when description is not null then description end ) as count_description
       ,count(case when listing_price is not null then listing_price end ) as count_listing_price
       ,count(case when last_visited is not null then last_visited end ) as count_last_visited
        
from info a 
left join finance b 
on a.product_id = b.product_id
left join reviews c 
on  a.product_id = c.product_id
left join traffic d 
on  a.product_id = d.product_id

; 




1 rows affected.


[(3179, 3117, 3120, 2928)]

## 2. Nike vs Adidas pricing
<p>We can see the database contains 3,179 products in total. Of the columns we previewed, only one &mdash; <code>last_visited</code> &mdash; is missing more than five percent of its values. Now let's turn our attention to pricing. </p>
<p>How do the price points of Nike and Adidas products differ? Answering this question can help us build a picture of the company's stock range and customer market. We will run a query to produce a distribution of the <code>listing_price</code> and the count for each price, grouped by <code>brand</code>. </p>

In [652]:
%%sql

select brand
    , listing_price :: int as listing_price
    , count(b.product_id) 
from brands a 
left join finance b 
on a.product_id = b.product_id
where listing_price>0
group by brand
    , listing_price 
order by listing_price desc

; 




 * postgresql:///sports
77 rows affected.


[('Adidas', 300, 2),
 ('Adidas', 280, 4),
 ('Adidas', 240, 5),
 ('Adidas', 230, 8),
 ('Adidas', 220, 11),
 ('Nike', 200, 1),
 ('Adidas', 200, 8),
 ('Nike', 190, 2),
 ('Adidas', 190, 7),
 ('Nike', 180, 4),
 ('Adidas', 180, 34),
 ('Nike', 170, 14),
 ('Adidas', 170, 27),
 ('Nike', 160, 31),
 ('Adidas', 160, 28),
 ('Adidas', 150, 41),
 ('Nike', 150, 6),
 ('Adidas', 140, 36),
 ('Nike', 140, 12),
 ('Adidas', 130, 96),
 ('Nike', 130, 12),
 ('Adidas', 120, 115),
 ('Nike', 120, 16),
 ('Nike', 110, 17),
 ('Adidas', 110, 91),
 ('Nike', 100, 14),
 ('Adidas', 100, 72),
 ('Adidas', 96, 2),
 ('Nike', 95, 1),
 ('Nike', 90, 13),
 ('Adidas', 90, 89),
 ('Adidas', 86, 7),
 ('Nike', 85, 5),
 ('Adidas', 85, 1),
 ('Adidas', 80, 322),
 ('Nike', 80, 16),
 ('Nike', 79, 1),
 ('Adidas', 76, 149),
 ('Adidas', 75, 1),
 ('Nike', 75, 7),
 ('Adidas', 70, 87),
 ('Nike', 70, 4),
 ('Adidas', 66, 102),
 ('Nike', 65, 1),
 ('Adidas', 63, 1),
 ('Nike', 60, 2),
 ('Adidas', 60, 211),
 ('Adidas', 56, 174),
 ('Adidas', 55, 2),
 ('Adidas', 53, 43),
 ('Nike', 50, 5),
 ('Adidas', 50, 183),
 ('Adidas', 48, 42),
 ('Nike', 48, 1),
 ('Adidas', 46, 163),
 ('Nike', 45, 3),
 ('Adidas', 45, 1),
 ('Adidas', 43, 51),
 ('Adidas', 40, 81),
 ('Nike', 40, 1),
 ('Adidas', 38, 24),
 ('Adidas', 36, 25),
 ('Adidas', 33, 24),
 ('Adidas', 30, 37),
 ('Nike', 30, 2),
 ('Adidas', 28, 38),
 ('Adidas', 27, 18),
 ('Adidas', 25, 28),
 ('Adidas', 23, 1),
 ('Adidas', 20, 8),
 ('Adidas', 18, 4),
 ('Adidas', 16, 4),
 ('Adidas', 15, 27),
 ('Adidas', 13, 27),
 ('Adidas', 12, 1),
 ('Adidas', 10, 11),
 ('Adidas', 9, 1)]

## 3. Labeling price ranges
<p>It turns out there are 77 unique prices for the products in our database, which makes the output of our last query quite difficult to analyze. </p>
<p>Let's build on our previous query by assigning labels to different price ranges, grouping by <code>brand</code> and <code>label</code>. We will also include the total <code>revenue</code> for each price range and <code>brand</code>. </p>

In [654]:
%%sql


select brand 
   ,count(b.product_id) 
    ,sum(revenue) as total_revenue
  ,case when listing_price < 42 then 'Budget' 
          when listing_price >=42 and listing_price< 74 then 'Average'
          when listing_price >=74 and listing_price< 129 then 'Expensive'
          when listing_price >= 129 then 'Elite'
    end as price_category
from brands a 
inner join finance b 
on a.product_id = b.product_id
where  a.brand is not null 
group by brand,case when listing_price < 42 then 'Budget' 
          when listing_price >=42 and listing_price< 74 then 'Average'
          when listing_price >=74 and listing_price< 129 then 'Expensive'
           when listing_price >= 129 then 'Elite'
    end
order by total_revenue desc

; 

 * postgresql:///sports
8 rows affected.


[('Adidas', 849, 4626980.069999999, 'Expensive'),
 ('Adidas', 1060, 3233661.060000001, 'Average'),
 ('Adidas', 307, 3014316.8299999987, 'Elite'),
 ('Adidas', 359, 651661.1200000002, 'Budget'),
 ('Nike', 357, 595341.0199999992, 'Budget'),
 ('Nike', 82, 128475.59000000003, 'Elite'),
 ('Nike', 90, 71843.15000000004, 'Expensive'),
 ('Nike', 16, 6623.5, 'Average')]

## 4. Average discount by brand
<p>Interestingly, grouping products by brand and price range allows us to see that Adidas items generate more total revenue regardless of price category! Specifically, <code>"Elite"</code> Adidas products priced \$129 or more typically generate the highest revenue, so the company can potentially increase revenue by shifting their stock to have a larger proportion of these products!</p>
<p>Note we have been looking at <code>listing_price</code> so far. The <code>listing_price</code> may not be the price that the product is ultimately sold for. To understand <code>revenue</code> better, let's take a look at the <code>discount</code>, which is the percent reduction in the <code>listing_price</code> when the product is actually sold. We would like to know whether there is a difference in the amount of <code>discount</code> offered between brands, as this could be influencing <code>revenue</code>.</p>

In [656]:
%%sql

select brand 
   ,avg(discount)*100 as average_discount
from brands a 
inner join finance b
on a.product_id = b.product_id
where  a.brand is not null 
group by brand
; 


 * postgresql:///sports
2 rows affected.


[('Nike', 0.0), ('Adidas', 33.452427184465606)]

## 5. Correlation between revenue and reviews
<p>Strangely, no <code>discount</code> is offered on Nike products! In comparison, not only do Adidas products generate the most revenue, but these products are also heavily discounted! </p>
<p>To improve revenue further, the company could try to reduce the amount of discount offered on Adidas products, and monitor sales volume to see if it remains stable. Alternatively, it could try offering a small discount on Nike products. This would reduce average revenue for these products, but may increase revenue overall if there is an increase in the volume of Nike products sold. </p>
<p>Now explore whether relationships exist between the columns in our database. We will check the strength and direction of a correlation between <code>revenue</code> and <code>reviews</code>. </p>

In [658]:
%%sql

select corr(reviews,revenue ) as review_revenue_corr
from reviews a 
inner join finance b 
on a.product_id =b.product_id


 * postgresql:///sports
1 rows affected.


[(0.6518512283481301,)]

## 6. Ratings and reviews by product description length
<p>Interestingly, there is a strong positive correlation between <code>revenue</code> and <code>reviews</code>. This means, potentially, if we can get more reviews on the company's website, it may increase sales of those items with a larger number of reviews. </p>
<p>Perhaps the length of a product's <code>description</code> might influence a product's <code>rating</code> and <code>reviews</code> &mdash; if so, the company can produce content guidelines for listing products on their website and test if this influences <code>revenue</code>. Let's check this out!</p>

In [660]:
%%sql

select TRUNC(length(description ),-2) as description_length 
       ,round(avg(rating :: numeric ),2) as average_rating
from info a 
inner join reviews b 
on a.product_id=b.product_id 
where description is not null
group by TRUNC(length(description ),-2) 
order by TRUNC(length(description ),-2)   




 * postgresql:///sports
7 rows affected.


[(Decimal('0'), Decimal('1.87')),
 (Decimal('100'), Decimal('3.21')),
 (Decimal('200'), Decimal('3.27')),
 (Decimal('300'), Decimal('3.29')),
 (Decimal('400'), Decimal('3.32')),
 (Decimal('500'), Decimal('3.12')),
 (Decimal('600'), Decimal('3.65'))]

## 7. Reviews by month and brand
<p>Unfortunately, there doesn't appear to be a clear pattern between the length of a product's <code>description</code> and its <code>rating</code>.</p>
<p>As we know a correlation exists between <code>reviews</code> and <code>revenue</code>, one approach the company could take is to run experiments with different sales processes encouraging more reviews from customers about their purchases, such as by offering a small discount on future purchases. </p>
<p>Let's take a look at the volume of <code>reviews</code> by month to see if there are any trends or gaps we can look to exploit.</p>

In [662]:
%%sql

select brand
        , extract(month from last_visited) as month 
        , count(c.product_id) as num_reviews
from brands a 
left join traffic b 
on a.product_id = b.product_id
left join reviews c 
on a.product_id = c.product_id
where brand is not null and  extract(month from last_visited) is not null
group by brand
        ,extract(month from last_visited)
order by brand, extract(month from last_visited)

 * postgresql:///sports
24 rows affected.


[('Adidas', Decimal('1'), 253),
 ('Adidas', Decimal('2'), 272),
 ('Adidas', Decimal('3'), 269),
 ('Adidas', Decimal('4'), 180),
 ('Adidas', Decimal('5'), 172),
 ('Adidas', Decimal('6'), 159),
 ('Adidas', Decimal('7'), 170),
 ('Adidas', Decimal('8'), 189),
 ('Adidas', Decimal('9'), 181),
 ('Adidas', Decimal('10'), 192),
 ('Adidas', Decimal('11'), 150),
 ('Adidas', Decimal('12'), 190),
 ('Nike', Decimal('1'), 52),
 ('Nike', Decimal('2'), 52),
 ('Nike', Decimal('3'), 55),
 ('Nike', Decimal('4'), 42),
 ('Nike', Decimal('5'), 41),
 ('Nike', Decimal('6'), 43),
 ('Nike', Decimal('7'), 37),
 ('Nike', Decimal('8'), 29),
 ('Nike', Decimal('9'), 28),
 ('Nike', Decimal('10'), 47),
 ('Nike', Decimal('11'), 38),
 ('Nike', Decimal('12'), 35)]

## 8. Footwear product performance
<p>Looks like product reviews are highest in the first quarter of the calendar year, so there is scope to run experiments aiming to increase the volume of reviews in the other nine months!</p>
<p>So far, we have been primarily analyzing Adidas vs Nike products. Now, let's switch our attention to the type of products being sold. As there are no labels for product type, we will create a Common Table Expression (CTE) that filters <code>description</code> for keywords, then use the results to find out how much of the company's stock consists of footwear products and the median <code>revenue</code> generated by these items.</p>

In [664]:
%%sql

with footwear as 
(
    select  a.product_id
            ,description
            ,revenue
    from info a 
    inner join finance b 
    on a.product_id = b.product_id 
    where (description ILIKE  '%shoe%' 
            or description ILIKE  '%trainer%' 
            or description ILIKE '%foot%')
            and description is not null
)

select COUNT( a.product_id) AS num_footwear_products
        , percentile_disc(0.5) WITHIN GROUP(ORDER BY revenue) as median_footwear_revenue
from  footwear 
;

 * postgresql:///sports
1 rows affected.


[(2700, 3118.36)]

## 9. Clothing product performance
<p>Recall from the first task that we found there are 3,117 products without missing values for <code>description</code>. Of those, 2,700 are footwear products, which accounts for around 85% of the company's stock. They also generate a median revenue of over $3000 dollars!</p>
<p>This is interesting, but we have no point of reference for whether footwear's <code>median_revenue</code> is good or bad compared to other products. So, for our final task, let's examine how this differs to clothing products. We will re-use <code>footwear</code>, adding a filter afterward to count the number of products and <code>median_revenue</code> of products that are not in <code>footwear</code>.</p>

In [666]:
%%sql


with footwear as 
(
    select  a.product_id
            ,description
            ,revenue
    from info a 
    inner join finance b 
    on a.product_id = b.product_id 
    where not (description ILIKE  '%shoe%' 
            or description ILIKE  '%trainer%' 
            or description ILIKE '%foot%')
            and description is not null
)

select COUNT(product_id) AS num_clothing_products
        , percentile_disc(0.5) WITHIN GROUP(ORDER BY revenue) as median_clothing_revenue
from  footwear 
;

 * postgresql:///sports
1 rows affected.


[(417, 503.82)]